### DISCUSSION TITLE
Language unique library features/changes

### DISCUSSION THREAD

**Discussion by thecoolwinter** [supabase/supabase#3153] [Feature Requests]
Languages often have different requirements or use cases depending on where they’re being built, and how objects are used in that language. I’m curious, is it okay for client libaries to make slight changes to the API or implementation of a client? 

Specifically, something like this PR: https://github.com/supabase/storage-swift/pull/1

In this case, swift devs often keeps objects alive for the entirety of an app’s lifecycle. Having to re-init an object with new headers could cause memory leaks and unexpected behavior. But implementing a config object, or even public config variables, splits from the precedent set by the JS and Dart libraries. 

Should this kind of change be okay? And I guess my suggestion would be: Can the client libraries follow the JS and Dart libraries as close as possible, but in the case where a change is necessary, keep API’s the same across the language?

**Comment by soedirgo** [comment]
Disclaimer: I haven't written a line of Swift and I'm not familiar with the Storage stuff.

> is it okay for client libaries to make slight changes to the API or implementation of a client?

I opined on this [here](https://github.com/supabase/supabase/discussions/5#discussioncomment-1159506) a while back - I'm open to the idea. But your statement sounded conflicting to me:

> swift devs often keeps objects alive for the entirety of an app’s lifecycle. Having to re-init an object with new headers could cause memory leaks and unexpected behavior

In that case, wouldn't you just init the object once and keep *that* alive? Do you have a use case for mutating an initialized client's headers/URL/etc.?

**Reply by soedirgo** [reply]
Also, it's not clear to me why initializing objects on demand would cause memory leaks. As long as you're not holding references to old objects, it should get cleaned up by the ARC.
